src/main/java/com/google/api/codegen/config/SampleSpec.java

```diff
@@ -46,16 +51,49 @@
     EXPLORER,
   }
 
+  /**
+   * A class used by {@code getMatchingValueSets} to return a value set needed to generate a sample
+   * for a given method, sample type, and calling form, together with region tag that should be used
+   * in that sample.
+   */
+  @AutoValue
+  public abstract static class ValueSetAndTags {
+
+    public abstract SampleValueSet values();
+
+    /** Returns the region tag prefix to be used for this sample. */
+    @Nullable
+    public abstract String regionTag();
+
+    public static Builder newBuilder() {
+      return new AutoValue_SampleSpec_ValueSetAndTags.Builder();
+    }
+
+    @AutoValue.Builder
+    public abstract static class Builder {
+
+      public abstract Builder values(SampleValueSet sets);
+
+      public abstract Builder regionTag(String tag);
+
+      public abstract ValueSetAndTags build();
+    }
+  }
+
   public SampleSpec(MethodConfigProto methodConfigProto) {
     specified = methodConfigProto.hasSamples();
     sampleConfiguration = methodConfigProto.getSamples();
     valueSets = methodConfigProto.getSampleValueSetsList();
 
-    HashSet<String> ids = new HashSet<>();
+    HashMap<String, SampleValueSet> setMap = new HashMap<>();
```

Comment:
I don't think this loop has to change at all. We're not actually using the `valueSet` themselves for anything here right?

Reply:
Good catch! Cruft from an earlier approach